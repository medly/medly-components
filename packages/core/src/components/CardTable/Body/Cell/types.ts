import { ColumnConfig, Data } from '../../types';

export interface Props {
    data: any;
    column: ColumnConfig;
    rowId: any;
    isRowClickDisabled?: boolean;
    rowData: Data;
}

export interface StyledProps {
    align?: 'left' | 'center' | 'right';
    wrapText?: boolean;
}
