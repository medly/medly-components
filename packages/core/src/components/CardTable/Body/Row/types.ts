import { WithThemeProp } from '@medly-components/utils';

export interface Props extends WithThemeProp {
    onClick?: () => void;
    isRowClickDisabled?: boolean;
    withWhiteBackground?: boolean;
    gridTemplateColumns: string;
}
