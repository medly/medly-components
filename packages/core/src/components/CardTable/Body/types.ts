import { ColumnConfig, Data } from '../types';

export interface Props {
    data: Data[];
    columns: ColumnConfig[];
    onRowClick?: (rowData: Record<string, unknown>) => void;
    rowClickDisableKey?: string;
    uniqueKeyName?: string;
    withWhiteBackground?: boolean;
}
