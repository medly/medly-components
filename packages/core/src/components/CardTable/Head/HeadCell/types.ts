import { ColumnConfig, SortOrder } from '../../types';

export interface Props {
    defaultSortOrder?: SortOrder | 'none';
    column: ColumnConfig;
    sortField?: string;
    withWhiteBackground?: boolean;
    onSortChange?: (field: string, order: SortOrder) => void;
}

export interface StyledProps {
    isActive?: boolean;
    sortable?: boolean;
    withWhiteBackground?: boolean;
    align?: 'left' | 'center' | 'right';
}
