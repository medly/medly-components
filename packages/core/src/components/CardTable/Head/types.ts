import { ColumnConfig, SortOrder } from '../types';

export interface Props {
    defaultSortField?: string;
    defaultSortOrder?: SortOrder;
    withWhiteBackground?: boolean;
    columns: ColumnConfig[];
    onSort?: (field: string, order: SortOrder) => void;
}
