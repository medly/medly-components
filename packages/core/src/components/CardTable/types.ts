import { HTMLProps } from '@medly-components/utils';
import type { FC } from 'react';

export type SortOrder = 'asc' | 'desc';

export interface Data {
    [key: string]: any;
}

export interface ColumnConfig {
    /** Title of the column */
    title: string;
    /** Field name in the data */
    field: string;
    /** Set it true to show sort icons */
    sortable?: boolean;
    /** Column content alignment */
    align?: 'left' | 'right' | 'center';
    /** Column Text content wrapping */
    wrapText?: boolean;
    /** Provide the fraction of the of the total width of the row */
    fraction?: number;
    /** Custom component */
    component?: FC<{ data: any; rowId?: any; disabled?: boolean; rowData?: Data }>;
}
export interface CardTableProps extends Omit<HTMLProps<HTMLOListElement>, 'data' | 'type'> {
    /** Table data */
    data: Array<{
        [key: string]: any;
    }>;
    /** Column configuration */
    columns: ColumnConfig[];
    /** Key name in data to be used as unique id */
    uniqueKeyName?: string;
    /** Default Sort Field*/
    defaultSortField?: string;
    /** Default Sort Order*/
    defaultSortOrder?: SortOrder;
    /** Function to be called on click of sort icon */
    onSort?: (field: string, order: SortOrder) => void;
    /** Function to be called on row click */
    onRowClick?: (rowData: Record<string, unknown>) => void;
    /** Key name to disable row click */
    rowClickDisableKey?: string;
    /** Key name to select background color */
    withWhiteBackground?: boolean;
}
