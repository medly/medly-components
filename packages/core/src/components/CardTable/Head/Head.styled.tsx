import styled from 'styled-components';

export const HeadRow = styled('tr')<{ gridTemplateColumns: string }>`
    display: grid;
    grid-template-columns: ${({ gridTemplateColumns }) => gridTemplateColumns};
    background-color: transparent;
`;
