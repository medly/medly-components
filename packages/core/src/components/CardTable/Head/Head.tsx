import React, { FC, useCallback, useMemo, useState } from 'react';
import { SortOrder } from '../types';
import * as Styled from './Head.styled';
import HeadCell from './HeadCell';
import { Props } from './types';

export const Head: FC<Props> = React.memo(({ onSort, columns, withWhiteBackground, defaultSortField, defaultSortOrder }) => {
    const [sortField, setSortField] = useState(defaultSortField);

    const gridTemplateColumns = useMemo(() => columns.reduce((acc, curr) => acc.concat(`${curr.fraction || 1}fr `), ''), [columns]),
        handleSortChange = useCallback(
            (field: string, order: SortOrder) => {
                setSortField(field);
                onSort && onSort(field, order);
            },
            [onSort]
        );

    return (
        <thead>
            <Styled.HeadRow gridTemplateColumns={gridTemplateColumns}>
                {columns.map(column => (
                    <HeadCell
                        defaultSortOrder={defaultSortOrder}
                        key={column.field}
                        sortField={sortField}
                        onSortChange={handleSortChange}
                        column={column}
                        withWhiteBackground={withWhiteBackground}
                    />
                ))}
            </Styled.HeadRow>
        </thead>
    );
});
Head.defaultProps = {
    defaultSortField: ''
};
Head.displayName = 'Head';
