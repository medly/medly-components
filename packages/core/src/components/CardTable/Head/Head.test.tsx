import { fireEvent, render } from '@test-utils';
import { ColumnConfig } from '../types';
import { Head } from './Head';

const patientColumns: ColumnConfig[] = [
        { title: 'Patient Info', field: 'patientInfo', sortable: true },
        { title: 'RxInfo', field: 'rxInfo', sortable: true }
    ],
    renderer = ({ withWhiteBackground = false, onSort = jest.fn(), columns = patientColumns }) =>
        render(
            <table>
                <Head onSort={onSort} columns={columns} withWhiteBackground={withWhiteBackground} />
            </table>
        );

describe('Card Table Head', () => {
    it('should render properly with dark background', () => {
        const { container } = renderer({ withWhiteBackground: false });
        expect(container).toMatchSnapshot();
    });

    it('should render properly with light background', () => {
        const { container } = renderer({ withWhiteBackground: true });
        expect(container).toMatchSnapshot();
    });

    it('should call onSort when clicked on sorting icon', () => {
        const mockOnSortChange = jest.fn(),
            { container } = renderer({ onSort: mockOnSortChange });

        fireEvent.click(container.querySelectorAll('svg')[0]);
        expect(mockOnSortChange).toHaveBeenCalledWith('patientInfo', 'asc');
    });
});
