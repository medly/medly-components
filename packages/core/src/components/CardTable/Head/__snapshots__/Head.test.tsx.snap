// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Table Head should render properly with dark background 1`] = `
.c0 {
  display: grid;
  grid-template-columns: 1fr 1fr;
  background-color: transparent;
}

.c7 {
  overflow: visible;
  font-size: 2rem;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  cursor: inherit;
}

.c7 * {
  fill-opacity: 1;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  fill: #607890;
}

.c5 {
  margin: 0;
  color: inherit;
  font-size: 1.2rem;
  font-weight: 700;
  -webkit-letter-spacing: 0.12rem;
  -moz-letter-spacing: 0.12rem;
  -ms-letter-spacing: 0.12rem;
  letter-spacing: 0.12rem;
  line-height: 2.0rem;
  text-align: initial;
  text-transform: uppercase;
}

.c3 {
  padding: 1rem 1.6rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  padding: 0 0.8rem;
  margin: 1.4rem 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c1::after {
  content: '';
  width: calc(100% - 1.6rem);
  margin: 0 0.8rem;
  display: block;
  position: absolute;
  bottom: -1.6rem;
  left: 0;
  background-color: transparent;
  height: 0.2rem;
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  border-radius: 0.2rem 0.2rem 0 0;
}

.c1 .c2:hover {
  cursor: pointer;
}

.c1 .c2:hover .c4 {
  color: #5687D9;
}

.c1 .c2:hover .c6 * {
  fill: #5687D9;
}

.c1 .c4 {
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline;
  color: #435465;
}

.c1 .c6 {
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  width: 2.4rem;
  cursor: pointer;
  padding-left: 0.4rem;
}

.c1 .c6 * {
  fill: #435465;
}

<div>
  <table>
    <thead>
      <tr
        class="c0"
      >
        <th
          class="c1"
        >
          <span
            class="c2 c3"
          >
            <h5
              class="c4 c5"
            >
              Patient Info
            </h5>
            <svg
              class="c6 c7"
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.229 6.935a1 1 0 0 1 1.543 0l2.208 2.679a1 1 0 0 1-.77 1.636H9.79a1 1 0 0 1-.77-1.636l2.209-2.679zM12.771 17.065a1 1 0 0 1-1.543 0L9.02 14.386a1 1 0 0 1 .772-1.636h4.418a1 1 0 0 1 .771 1.636l-2.209 2.679z"
                fill="#435465"
              />
            </svg>
          </span>
        </th>
        <th
          class="c1"
        >
          <span
            class="c2 c3"
          >
            <h5
              class="c4 c5"
            >
              RxInfo
            </h5>
            <svg
              class="c6 c7"
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.229 6.935a1 1 0 0 1 1.543 0l2.208 2.679a1 1 0 0 1-.77 1.636H9.79a1 1 0 0 1-.77-1.636l2.209-2.679zM12.771 17.065a1 1 0 0 1-1.543 0L9.02 14.386a1 1 0 0 1 .772-1.636h4.418a1 1 0 0 1 .771 1.636l-2.209 2.679z"
                fill="#435465"
              />
            </svg>
          </span>
        </th>
      </tr>
    </thead>
  </table>
</div>
`;

exports[`Card Table Head should render properly with light background 1`] = `
.c0 {
  display: grid;
  grid-template-columns: 1fr 1fr;
  background-color: transparent;
}

.c7 {
  overflow: visible;
  font-size: 2rem;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  cursor: inherit;
}

.c7 * {
  fill-opacity: 1;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  fill: #607890;
}

.c5 {
  margin: 0;
  color: inherit;
  font-size: 1.2rem;
  font-weight: 700;
  -webkit-letter-spacing: 0.12rem;
  -moz-letter-spacing: 0.12rem;
  -ms-letter-spacing: 0.12rem;
  letter-spacing: 0.12rem;
  line-height: 2.0rem;
  text-align: initial;
  text-transform: uppercase;
}

.c3 {
  padding: 1rem 1.6rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  padding: 0 0.8rem;
  margin: 1.4rem 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c1::after {
  content: '';
  width: calc(100% - 1.6rem);
  margin: 0 0.8rem;
  display: block;
  position: absolute;
  bottom: -1.6rem;
  left: 0;
  background-color: transparent;
  height: 0.2rem;
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  border-radius: 0.2rem 0.2rem 0 0;
}

.c1 .c2:hover {
  cursor: pointer;
}

.c1 .c2:hover .c4 {
  color: #5687D9;
}

.c1 .c2:hover .c6 * {
  fill: #5687D9;
}

.c1 .c4 {
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline;
  color: #607890;
}

.c1 .c6 {
  -webkit-transition: all 100ms ease-out;
  transition: all 100ms ease-out;
  width: 2.4rem;
  cursor: pointer;
  padding-left: 0.4rem;
}

.c1 .c6 * {
  fill: #607890;
}

<div>
  <table>
    <thead>
      <tr
        class="c0"
      >
        <th
          class="c1"
        >
          <span
            class="c2 c3"
          >
            <h5
              class="c4 c5"
            >
              Patient Info
            </h5>
            <svg
              class="c6 c7"
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.229 6.935a1 1 0 0 1 1.543 0l2.208 2.679a1 1 0 0 1-.77 1.636H9.79a1 1 0 0 1-.77-1.636l2.209-2.679zM12.771 17.065a1 1 0 0 1-1.543 0L9.02 14.386a1 1 0 0 1 .772-1.636h4.418a1 1 0 0 1 .771 1.636l-2.209 2.679z"
                fill="#435465"
              />
            </svg>
          </span>
        </th>
        <th
          class="c1"
        >
          <span
            class="c2 c3"
          >
            <h5
              class="c4 c5"
            >
              RxInfo
            </h5>
            <svg
              class="c6 c7"
              fill="none"
              height="1em"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.229 6.935a1 1 0 0 1 1.543 0l2.208 2.679a1 1 0 0 1-.77 1.636H9.79a1 1 0 0 1-.77-1.636l2.209-2.679zM12.771 17.065a1 1 0 0 1-1.543 0L9.02 14.386a1 1 0 0 1 .772-1.636h4.418a1 1 0 0 1 .771 1.636l-2.209 2.679z"
                fill="#435465"
              />
            </svg>
          </span>
        </th>
      </tr>
    </thead>
  </table>
</div>
`;
